### PR TITLE
add some args (upload/download interval and parallelism on upload)

### LINKED ISSUE #3: Sleeps make tool slower than it needs to be?
Hi, great tool.

It's not entirely clear what the purpose of these sleeps are:

https://github.com/netmikey/mvncloner/blob/master/src/main/java/io/github/netmikey/mvncloner/mvncloner/Scraper.java#L133
https://github.com/netmikey/mvncloner/blob/master/src/main/java/io/github/netmikey/mvncloner/mvncloner/Scraper.java#L153
https://github.com/netmikey/mvncloner/blob/master/src/main/java/io/github/netmikey/mvncloner/mvncloner/Publisher.java#L83

The tool seems much faster without them (I've built and tested it). Is there some issue that they're pre-empting?

### PR DESCRIPTION
**Pull request by cleciusjm**
Created some args to allow more personalized usage:
* dry-run: to test all args without any write action on target
* target.upload-interval: allow to customize delay between uploads, useful on internal servers when load could be bigger
* target.concurrent-uploads: allow concurrent uploads in same folder, useful as previous
* source.download-interval: allow to customize delay between downloads, also useful on internal servers

### CODE REVIEW

File: src/main/java/io/github/netmikey/mvncloner/mvncloner/Publisher.java [hunk 1]
```diff
@@ -39,63 +39,86 @@ public class Publisher {
...
     private String password;
 
+    @Value("${dry-run:false}")
+    private Boolean dryRun;
+
+    @Value("${target.upload-interval:1000}")
+    private Integer uploadInterval;
+
+	@Value("${target.concurrent-uploads:2}")
+	private Integer concurrentUploads;
```
Comment (netmikey) [inline src/main/java/io/github/netmikey/mvncloner/mvncloner/Publisher.java]: It seems you have introduced a mix of space and tab indentations. Please stick to the convention used throughout the project (indentation using 4 spaces).

File: src/main/java/io/github/netmikey/mvncloner/mvncloner/Publisher.java [hunk 2]
```diff
@@ -39,63 +39,86 @@ public class Publisher {
...
+            var putRequest = baseReq.PUT(BodyPublishers.ofFile(path)).build();
+
+            HttpResponse<String> putResponse = httpClient.send(putRequest, BodyHandlers.ofString());
+            if (putResponse.statusCode() < 200 || putResponse.statusCode() > 299) {
+                LOG.error("Error uploading " + targetUrl + " : Response code was " + putResponse.statusCode());
+                LOG.debug("   Response headers: " + putResponse.headers());
+                LOG.debug("   Response body: " + putResponse.body());
+            }
+        } catch (Exception e) {
+            LOG.error("Fail to send " + filename + " to " + targetUrl, e);
```
Comment (netmikey) [inline src/main/java/io/github/netmikey/mvncloner/mvncloner/Publisher.java]: This changes the behavior quite a bit: Before, the Publisher would abort on `IOException` and `InterruptedException` (which probably won't be recoverable anyway, so no use to continue to try uploading). Now, you just log and continue uploading.

I'd rather not change the general error behavior, so if you'd like to introduce concurrency, please handle errors accordingly.

File: src/main/java/io/github/netmikey/mvncloner/mvncloner/Publisher.java [hunk 3]
```diff
@@ -39,63 +39,86 @@ public class Publisher {
...
+        try {
+            var baseReq = Utils.setCredentials(HttpRequest.newBuilder(), username, password)
+                .uri(URI.create(targetUrl));
+
+            var getResponse = httpClient.send(baseReq.method("HEAD", BodyPublishers.noBody()).build(), BodyHandlers.discarding());
+            if (getResponse.statusCode() != 404) {
+                LOG.info("Artifact {} already exists", targetUrl);
+                LOG.debug("   Response headers: " + getResponse.headers());
+                return;
+            }
```
Comment (netmikey) [inline src/main/java/io/github/netmikey/mvncloner/mvncloner/Publisher.java]: Why did you add an additional http HEAD request? Not sure if this is needed and  backwards compatible? Also, assuming anything but http 404 means the file already exists is probably not correct.

File: README.md [hunk 4]
```diff
@@ -31,22 +31,26 @@ Each argument is prepended with "`--`".
...
+source.download-interval | No       | The interval between downloads, value in milliseconds, used to reduce load on source-server (default: 1000) 
+mirror-path              | No       | The path on the local file system to be used for mirroring repository content (default: `./mirror/`)
+target.root-url          | Yes      | The root URL where the content of the mirror repository should be uploaded.
+target.user              | No       | The username used on the target repository (default: no authentication).
+target.password          | No       | The password used on the target repository (default: no authentication).
+target.upload-interval   | No       | The interval between uploads, value in milliseconds, used to reduce load on target-server (default: 1000)
+target.concurrent-uploads| No       | Number of parallel uploads, parallelism is only supported on same folder, like foo/bar/artifact/1.0.0/ (default: 2)
+dry-run                  | No       | Boolean value (true or false) that disable write action, but run all read avaliable steps, useful when connections need to be checked without any change (default: false)
```
Comment (netmikey) [inline README.md]: I thought about adding a `dry-run` option but chose the `actions` mechanism below because it is clearer what the tool actually does. I think the `dry-run` option now collides with the semantics of the `actions` option. Using `--actions=mirror` basically does the same (minus the http HEAD request, see below). 

Also I would expect a true "dry-run" to also **not** perform the mirror action, because that writes on the local filesystem and hence isn't really dry.

I'd prefer sticking to the `actions` option only. If performing a "connection-check" is your goal, I'd suggest adding a "check" action that does exactly this on the source and the target but nothing else.

File: README.md [hunk 5]
```diff
@@ -31,22 +31,26 @@ Each argument is prepended with "`--`".
...
+source.user              | No       | The username used on the source repository (default: no authentication).
+source.password          | No       | The password used on the source repository (default: no authentication).
+source.download-interval | No       | The interval between downloads, value in milliseconds, used to reduce load on source-server (default: 1000) 
+mirror-path              | No       | The path on the local file system to be used for mirroring repository content (default: `./mirror/`)
+target.root-url          | Yes      | The root URL where the content of the mirror repository should be uploaded.
+target.user              | No       | The username used on the target repository (default: no authentication).
+target.password          | No       | The password used on the target repository (default: no authentication).
+target.upload-interval   | No       | The interval between uploads, value in milliseconds, used to reduce load on target-server (default: 1000)
+target.concurrent-uploads| No       | Number of parallel uploads, parallelism is only supported on same folder, like foo/bar/artifact/1.0.0/ (default: 2)
```
Comment (netmikey) [inline README.md:51]: This might be nitpicking but one might ask: What is the reason for adding concurrent uploading but not concurrent downloading?

**Review by netmikey — CHANGES_REQUESTED** [review]

File: src/main/java/io/github/netmikey/mvncloner/mvncloner/Publisher.java (hunk 2 above)
Comment (cleciusjm) [reply src/main/java/io/github/netmikey/mvncloner/mvncloner/Publisher.java]: This change is intetional, i've some problems with high loads or networks instabilities, that is why a idempotent way of submit has implemented too, to allow some fails to be send with lower impact

File: src/main/java/io/github/netmikey/mvncloner/mvncloner/Publisher.java (hunk 3 above)
Comment (cleciusjm) [reply src/main/java/io/github/netmikey/mvncloner/mvncloner/Publisher.java]: This is to be more idempotent, in large process some fails occours (here about 3 days of sync) and re-upload all data only to get a fail isn't fast nor efficient. 
with this, if the file already exists, the upload no longer is executed

File: README.md (hunk 4 above)
Comment (cleciusjm) [reply README.md]: i agree with some of this, but, the target check isn't validated without starting to write without this dry-run action

File: README.md (hunk 5 above)
Comment (cleciusjm) [reply README.md:51]: in my scenario, download is fron an old server who does'nt have a high bandwith, but upload is to a new server who benefits from parallelism, and with idempotency, high amount of head requests could be done concurrently when the processes is interrupted and restarted, until get a valid new upload

File: src/main/java/io/github/netmikey/mvncloner/mvncloner/Publisher.java (hunk 2 above)
Comment (netmikey) [reply src/main/java/io/github/netmikey/mvncloner/mvncloner/Publisher.java]: Ok I see.

This seems legit for `IOException` then, but the tool should definitely always abort on `InterruptedException`.

Thinking further here: I'd like to avoid having to scrape the logs at the end of a long upload session to find if everything worked or not. If the tool aborts at the first error, it is clear where / when things went wrong but if the tool continues for a long time, errors might get overseen.

I can think of 2 things to help solve this:

1. Make it configurable: something like `--target.abort-on-error=false`. This way the old behavior can be restored for people on stable networks that might prefer the current behavior.
2. Count the number of failed uploads and print a summary line at the very end (in addition to the error logs). This way, if `--target.abort-on-error=false` is set, you can be sure that nothing failed, or how much failed during the long-running operation.

File: src/main/java/io/github/netmikey/mvncloner/mvncloner/Publisher.java (hunk 3 above)
Comment (netmikey) [reply src/main/java/io/github/netmikey/mvncloner/mvncloner/Publisher.java]: Ok I see.

Would you mind making it configurable though? I could imagine that there's a use-case out there where people want to explicitly overwrite all or parts of an already existing repository. In such case, people could cut the amount of http requests in half by omitting the http HEAD request and PUT'ting immediately.

Something like: `--target.skip-existing=true`.

File: README.md (hunk 4 above)
Comment (netmikey) [reply README.md]: I'm not sure I understand what you mean here. My point is: rather than adding a `dry-run` parameter, capsule whatever is necessary for the connection-checking into a new `check` action.

I'd guess that a http HEAD request for the `target.root-url` would be enough to validate the target credentials isn't it?

File: README.md (hunk 5 above)
Comment (netmikey) [reply README.md:51]: Ok I see, thanks for the clarification. :-)

File: src/main/java/io/github/netmikey/mvncloner/mvncloner/Publisher.java (hunk 2 above)
Comment (cleciusjm) [reply src/main/java/io/github/netmikey/mvncloner/mvncloner/Publisher.java]: collect erros recursively could be more complex, we need to think more about lifecycle of Publisher instance, usually we re-run to check if some fail has ocurred.

But the flag is a good thing, i'll add it

**Comment by netmikey** [comment]
@cleciusjm I added a couple of improvements according to what we discussed above. I don't have test Nexus servers at hand anymore unfortunately. Would you mind testing my changes to see if they work as described in the README and don't break existing behavior?

**Comment by cleciusjm** [comment]
nice, only check action feel like duplicated code to me, but ok

my new nexus is in production, i cant run publish action, but mirror and check seems ok to validate here

**Comment by netmikey** [comment]
I agree about the code duplication. The 3 lines of `HttpClient` code felt too trivial to extract and the `WebClient` API with its try-with-resource would've been cumbersome to make DRY...

Anyway, let me know when you've been able to test-drive the mirror and check actions :-)

**Comment by netmikey** [comment]
nvm, I got rid of the duplicated `WebClient` code.

**Comment by cleciusjm** [comment]
Check is failing because target address itself resolves as 404, but internal artifact addresses prefixed by it are funcional.


```
2023-08-23 09:14:38.715  INFO 16841 --- [           main] i.g.n.mvncloner.mvncloner.Checker        : Source connection check succeeded! (http://xxxxxx/nexus/service/rest/repository/browse/releases/, responded with 200)
2023-08-23 09:14:39.547  INFO 16841 --- [           main] ConditionEvaluationReportLoggingListener : 

Error starting ApplicationContext. To display the conditions report re-run your application with 'debug' enabled.
2023-08-23 09:14:39.552 ERROR 16841 --- [           main] o.s.boot.SpringApplication               : Application run failed

java.lang.IllegalStateException: Failed to execute CommandLineRunner
	at org.springframework.boot.SpringApplication.callRunner(SpringApplication.java:798) ~[spring-boot-2.3.1.RELEASE.jar:2.3.1.RELEASE]
	at org.springframework.boot.SpringApplication.callRunners(SpringApplication.java:779) ~[spring-boot-2.3.1.RELEASE.jar:2.3.1.RELEASE]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:322) ~[spring-boot-2.3.1.RELEASE.jar:2.3.1.RELEASE]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1237) ~[spring-boot-2.3.1.RELEASE.jar:2.3.1.RELEASE]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1226) ~[spring-boot-2.3.1.RELEASE.jar:2.3.1.RELEASE]
	at io.github.netmikey.mvncloner.mvncloner.MvnclonerApplication.main(MvnclonerApplication.java:10) ~[main/:na]
Caused by: java.lang.IllegalStateException: Connection check failed: target at https://xxxxxx/nexus/repository/releases/ responded with http status code 404
	at io.github.netmikey.mvncloner.mvncloner.Checker.checkTarget(Checker.java:89) ~[main/:na]
	at io.github.netmikey.mvncloner.mvncloner.Checker.check(Checker.java:53) ~[main/:na]
	at io.github.netmikey.mvncloner.mvncloner.MvnCloner.run(MvnCloner.java:33) ~[main/:na]
	at org.springframework.boot.SpringApplication.callRunner(SpringApplication.java:795) ~[spring-boot-2.3.1.RELEASE.jar:2.3.1.RELEASE]
	... 5 common frames omitted

```

**Comment by cleciusjm** [comment]
that's why my previus check action was targeting artifacts from mirror, and check everything working status without only uploading new data

**Comment by netmikey** [comment]
> Check is failing because target address itself resolves as 404, but internal artifact addresses prefixed by it are funcional.
> that's why my previus check action was targeting artifacts from mirror, and check everything working status without only uploading new data

Well, if you ran your initial code against an empty target repository, this repository would've also returned 404 since you were doing an http HEAD on the target artifacts themselves, which aren't present on an empty repository. The thing is that [you accepted http 404 as the expected return code and went on](https://github.com/netmikey/mvncloner/pull/5/commits/92fae6b45e4f225e798765c495a328a736d072fd#diff-030385ca63874f6c40dd5eeef6919b0f353cad7a6ca839a9833574710f60bbbfR102-R110).

So in short, I think without writing to the target there's nothing else we can do but check that we get an http 200 (or 404) in response to the root URL, which would be the only existing URL in an empty repository.

I added a commit that accepts http 404 now too for the target check.
